lib/auth.js

```diff
@@ -11,65 +11,116 @@ const mkdirp = pify(require('mkdirp'));
+        cfg.bmp = cfg.bmp || {};
+        const origin = urlUtils.getOrigin(projectScope);
+        cfg.bmp[origin] = {
+          credential: options.credential
+        };
+        return cfg;
+      }, {
+        mode: 0o600
       });
-  },
+    });
+}
 
-  logout (options) {
-    options = options || {};
-    if (!options.scope) {
-      return Promise.reject(new Error('must set project scope first'));
-    }
-    const userConfigFile = values.userConfigFile(options.userConfigDir);
-    const EXITEARLYHACK = 'EXITEARLYHACK';
-    return fsp.access(userConfigFile, fs.F_OK)
-      .catch(() => {
-        return Promise.reject(new Error('EXITEARLYHACK'));
-      })
-      .then(() => fsp.access(userConfigFile, fs.R_OK | fs.W_OK))
-      .then(() => {
-        return jsonUtils.updateJson(userConfigFile, (cfg) => {
-          cfg.bmp = cfg.bmp || {};
-          if (cfg.bmp[getOrigin(options.scope)]) {
-            delete cfg.bmp[getOrigin(options.scope)].credential;
-          }
-          return cfg;
-        }, {
-          mode: 0o600
-        });
-      })
-      .catch((err) => {
-        if (err.message !== EXITEARLYHACK) {
-          return Promise.reject(err);
+function logout () {
+  const EXITEARLYHACK = 'EXITEARLYHACK';
+
+  let projectScope;
+  return scope.read()
+    .then((s) => {
+      projectScope = s;
+    })
+    .catch(() => {
+      throw new Error('must set project scope first');
+    })
+    .then(() => fsp.access(values.USER_CONFIG_FILE, fs.F_OK))
+    .catch(() => {
+      return Promise.reject(new Error('EXITEARLYHACK'));
+    })
+    .then(() => fsp.access(values.USER_CONFIG_FILE, fs.R_OK | fs.W_OK))
+    .then(() => {
+      return jsonUtils.updateJson(values.USER_CONFIG_FILE, (cfg) => {
+        cfg.bmp = cfg.bmp || {};
+        const origin = urlUtils.getOrigin(projectScope);
+        if (cfg.bmp[origin]) {
+          delete cfg.bmp[origin].credential;
         }
+        return cfg;
+      }, {
+        mode: 0o600
       });
-  }
+    })
+    .catch((err) => {
+      if (err.message !== EXITEARLYHACK) {
+        return Promise.reject(err);
+      }
+    });
+}
 
+function readAll () {
+  return jsonUtils.loadJsonObject(values.USER_CONFIG_FILE)
+    .catch((err) => {
+      if (err.code === 'ENOENT') {
+        // missing file = not authenticated, which we handle further down
+        return {};
+      }
+      // throw every other error, e.g. corrupt file, bad permissions, etc
+      throw err;
+    })
+    .then((obj) => {
+      const bmp = obj.bmp || {};
+      return Object.keys(bmp)
+        .filter((origin) => !!bmp[origin].credential)
+        .map((origin) => {
+          const credential = bmp[origin].credential;
+          return { origin, credential };
+        });
+    });
+}
+
+function read () {
+  let origin;
+  return scope.read()
+    .then((projectScope) => {
+      origin = urlUtils.getOrigin(urlUtils.toHTTPS(projectScope) || '');
+    })
+    .then(() => readAll())
```

Comment:
i think there's no need for a new function here, i think this should achieve the same thing
    
    .then(readAll)